### PR TITLE
Add Binary search recur in C code

### PR DESCRIPTION
**Pull request by Zuoxun**
Running results are as follows:

> 目标元素 6 的索引 = 2

If this PR is related to coding or code translation, please fill out the checklist and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

**Comment by Zuoxun** [comment]
Code binary_search_recur.c would rely on utils/vector.h, so this PR has already included the content of PR “Update vector.h” .

File: codes/c/chapter_divide_and_conquer/CMakeLists.txt
Comment (Gonglja) [inline codes/c/chapter_divide_and_conquer/CMakeLists.txt:1]: Remove add_executable syntax errors that are not related to this PR. For details, see #819

File: codes/c/utils/vector.h
Comment (Gonglja) [inline codes/c/utils/vector.h:1]: https://github.com/krahets/hello-algo/pull/819#discussion_r1346139561

File: codes/c/chapter_divide_and_conquer/binary_search_recur.c
Comment (Gonglja) [inline codes/c/chapter_divide_and_conquer/binary_search_recur.c:1]: Please add execution results

File: codes/c/chapter_divide_and_conquer/binary_search_recur.c [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    int n = nums->size;
+    // 求解问题 f(0, n-1)
+    return dfs(nums, target, 0, n - 1);
+}
+
+/* Driver Code */
+int main() {
+    int target = 6;
+    int nums[] = {1, 3, 6, 8, 12, 15, 23, 26, 31, 35};
+    vector *iNums = newVector(); // int
```
Comment (Gonglja) [inline codes/c/chapter_divide_and_conquer/binary_search_recur.c:41]: Memory not released

File: codes/c/chapter_divide_and_conquer/binary_search_recur.c
Comment (Zuoxun) [reply codes/c/chapter_divide_and_conquer/binary_search_recur.c:1]: Running results are as follows:

> 目标元素 6 的索引 = 2

**Review by Gonglja — COMMENTED** [review]
Can be approved @krahets

File: codes/c/chapter_divide_and_conquer/binary_search_recur.c [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
```
Comment (krahets) [inline codes/c/chapter_divide_and_conquer/binary_search_recur.c:41]: Why do you replace array with vector?

Prefer to use array in C when there is no add/remove operations.

Comment (Zuoxun) [reply codes/c/chapter_divide_and_conquer/binary_search_recur.c]: True, i've replaced vector with array now. Please review again.

And new running result:
> 目标元素 6 的索引 = 2

**Review by krahets — APPROVED** [review]
Thanks for the PR!
